File: Erlang/0010.erl
-export([is_match/2]).

-spec is_match(S :: unicode:unicode_binary(), P :: unicode:unicode_binary()) -> boolean().
is_match(S, P) ->
    Memo = ets:new(memo, [set, public, named_table]),
    Result = is_match_helper(S, P, 0, 0, byte_size(S), byte_size(P), Memo),
    ets:delete(Memo),
    Result.

-spec is_match_helper(S :: unicode:unicode_binary(), P :: unicode:unicode_binary(), I :: integer(), J :: integer(), SLen :: integer(), PLen :: integer(), Memo :: ets:tid()) -> boolean().
is_match_helper(_, _, I, J, SLen, PLen, _) when J == PLen ->
    I == SLen;
is_match_helper(S, P, I, J, SLen, PLen, Memo) ->
    case ets:lookup(Memo, {I, J}) of
        [{_, Result}] -> Result;
        [] ->
            FirstMatch = I < SLen andalso
                         (binary:at(S, I) =:= binary:at(P, J) orelse binary:at(P, J) =:= $.),
            IsStar = J + 1 < PLen andalso is_star(P, J + 1),
            Result = if
                IsStar ->
                    is_match_helper(S, P, I, J + 2, SLen, PLen, Memo) orelse
                    (FirstMatch andalso is_match_helper(S, P, I + 1, J, SLen, PLen, Memo));
                true ->
                    FirstMatch andalso is_match_helper(S, P, I + 1, J + 1, SLen, PLen, Memo)
            end,
            ets:insert(Memo, {{I, J}, Result}),
            Result
    end.

is_star(P, Index) ->
    binary:at(P, Index) =:= $*.

File: Elixir/0010.exs
defmodule Solution do
  @spec is_match(s :: String.t(), p :: String.t()) :: boolean
  def is_match(s, p) do
    memo = :ets.new(:memo, [:set, :public, :named_table])
    result = is_match_helper(s, p, 0, 0, String.length(s), String.length(p), memo)
    :ets.delete(:memo)
    result
  end

  defp is_match_helper(s, p, i, j, s_len, p_len, memo) when j == p_len do
    i == s_len
  end

  defp is_match_helper(s, p, i, j, s_len, p_len, memo) do
    case :ets.lookup(memo, {i, j}) do
      [{_, result}] -> result
      [] ->
        first_match = i < s_len and (String.at(s, i) == String.at(p, j) or String.at(p, j) == ".")

        result =
          if j + 1 < p_len and String.at(p, j + 1) == "*" do
            is_match_helper(s, p, i, j + 2, s_len, p_len, memo) or
              (first_match and is_match_helper(s, p, i + 1, j, s_len, p_len, memo))
          else
            first_match and is_match_helper(s, p, i + 1, j + 1, s_len, p_len, memo)
          end

        :ets.insert(memo, {{i, j}, result})
        result
    end
  end
end
